### ISSUE TITLE
Inventory database

### ISSUE THREAD

**Issue by foxfriends** [TerrenceJChan/discord-rpg-game#26]
The player can own items. Items are shared between all characters a player has created. 

There are three main types of item:
1. Currency
2. Material
3. Equipment

Each piece of equipment has its own name and modifiers, so they are tracked individually. Currencies and materials, on the other hand, are basically the same and can stack (infinitely?).

---

Maybe add something like these tables (primary key in `{}`)?

* `Materials ({player_id, material_id}, quantity)`: materials and quantity
* `Equipment ({id}, player_id, equipment_id, equipment_type, name, base_stats...)`: list of equipment
* `Modifiers ({id}, equiment_id, name, stat_modifier...)`: one row for each modifier on a piece of equipment (e.g. one equipment has 3 rows in modifiers)

Notably the player's actual inventory is spread out over two tables. We'll probably do some stuff in the code to hide this distinction from the users as much as possible, showing them just "inventory". Often times database does not line up 100% exactly with the things the users see or even the things we consider to be the same in code, and that's ok. The database doesn't care about the code or anything, it just wants valid data. If we make no sacrifices on representing the data as accurately as possible, the code often turns out best anyway.

As for the field values, my only real suggestions:
* `material_id` is probably an `enum_material`, similar to enemy. 
* `equipment_type` also an `enum_equipment_type` with things like `Weapon`, `Armor`, `Shield`, etc, whatever equipment types you are imagining.
* `equipment_id` can probably be yet another `enum_equipment`, which knows all of the different types of weapons/armor/etc we will invent. 

The one downside to using enums for all those things is that each time you add new materials or weapons, you'll need to update the enum in the database also. If this proves to be too much of a pain, we can easily convert the enum to `VARCHAR` and use that instead, but we lose the validation that enum provides about only storing valid items. We start with enum though because __*it's always easier to make things more flexible than to make them stricter, so we start with the strictest, even if it is a pain, and loosen when it hurts*__.

The big questions are:
1. What does `base_stats` look like?
2. What is a `stat_modifier`?

**Comment by TerrenceJChan** [comment]
`base_stats` are going to be a constant across each weapon type. For example, every `DragonsBite` weapon will have the same base stats across the board (not including upgrade, but that's a later problem). This can be handled in the front-end.

`stat_modifier` is likely where we will find some interesting challenges pop up. As I mentioned before, each weapon will have a pool of stat modifiers with a range of values associated with each mod. We'll have to save them this information in the database since this will be unique to each weapon instance.

**Comment by TerrenceJChan** [comment]
`equipment_type` as an enum makes sense. We'll have the slots like you mention (`Weapon`, `Armor`, etc.), but materials and equipment could theoretically number in tens or hundreds when things get going. Following your advice though, we'll start them out as enums so we can't forcefully jam things into them until we absolutely need to.

**Comment by foxfriends** [comment]
Nothing wrong with an enum with hundreds of options - still more efficient and more reliable than strings... but also chance of missing adding one to the list, which would cause problem. We'll see how it goes, maybe there can be a check on another GitHub action to make sure we never forget one

**Comment by TerrenceJChan** [comment]
Okay, so thinking through the stat modifiers, suppose we have a pool of 100 different mods across the entire game. Let's say that our `DragonsBite` weapon has 4 of those mods. Should we have a `stat_modifier` table that has 100 columns, linked to our `equipment` table that only has four cells with values? Is that the best way to go about it?

**Comment by foxfriends** [comment]
Maybe just another enum_stat to decide which stat to affect? Or will some modifiers affect multiple things?

**Comment by TerrenceJChan** [comment]
Enums might be the way to go I guess. Is it possible for a table to contain multiples of the same enum type?

**Comment by foxfriends** [comment]
Yeah that would be no problem, just don't put `UNIQUE` on the column and it'll be fine

**Comment by TerrenceJChan** [comment]
How would that actually look on a table? Would it be one cell with a list of modifiers?

In games like these, special items usually have special modifiers that are exclusive to them. If only one item has a specific mod, I assume it's still not an issue due to the nature of enums, right?

**Comment by foxfriends** [comment]
We can work it out sometime this weekend I guess. Need to decide exactly what is a modifier, what kinds of things they can affect, how do they work etc.

1. Can modifiers affect more than one stat?
2. Do they always work by adjusting one number, or are there some that do something else?
3. Are the modifiers shared from a list, will any be generated dynamically, or specially for a type of weapon?

I guess also worth deciding is how much to store in the DB about each weapon... if I had a `Long Sword`, and we were rebalancing the game, do we need to update the previously obtained Long Sword, or do we just add a new long sword and let the old one die off or be upgraded or something? 

Another way we could store modifiers, if they are purely a game thing and don't need to be validated by the database too much is `JSONB`: store any JSON object in the database directly. Kinda like Mongo, but inside a Postgres. This may be a situation where that is desired, as it gives a lot of flexibility, just also requires a bit more careful usage. I'm kinda starting to think for this problem that might be the best solution?

**Comment by TerrenceJChan** [comment]
1. Modifiers generally affect only one stat. It may be possible that we have different elemental attacks, for example, so we'll have a "Fire Attack" stat/mod, "Frost Attack" stat/mod, and a "Shock Attack" stat/mod. It's conceivable for a weapon to have the mods "Fire Attack +X%", and possibly "Elemental Attack +X%", where two different mods can effect the same stat, or one stat giving a general increase to all the elemental attacks.
2. Generally, mods affect only one number. That said, if we want to take it a step further where "unique" mods can make certain items special, it's possible that they'll come along with special conditions and effects. For example, there could be a pair of special Gloves that increases the user's attack by 300% if they don't equip a weapon, or maybe Armor that prevents them from being frozen (status effect).
3. There would be a long list of mods that exist overall. Each item will have access to certain mods from that pool. So if there are 100 mods overall, our `DragonsBite` weapon might have access to 20 of them (it could have access to mods that increase fire attacks, but not frost attacks for example). Depending on the item level/tier of the item in question, there would be different ranges of possible rolls. A low level `DragonsBite` might have access to "Fire Attack +X%" with a range of 8%-15%, but an upgraded `DragonsBite` might have the same mod but with a range of 24%-31%.

We can go over this in a call if you'd like. Stuff like this is very common in ARPGs but can seem very excessive. We can break down how other games do this and see what's happening under the hood.

You're also going to have to explain to me how we could use `JSONB` in more detail haha. Are you suggesting that we have, say, a column for mods where the cell takes a JSON object?

**Comment by TerrenceJChan** [comment]
I had a quick look at `JSONB` types, and yes, this does look how I envision the data being stored too. The DB just needs to know that this particular `DragonsBite` has mods A, B, C, D and the numeric value of each of them. The front-end code can handle the rest.
